File: src/main/java/dev/alexengrig/leetcode/_36_valid_sudoku/SetSolution.java
package dev.alexengrig.leetcode._36_valid_sudoku;

import java.util.HashSet;
import java.util.Set;

public class SetSolution implements Task {
    @Override
    public boolean isValidSudoku(char[][] board) {
        int n = board.length;
        int m = (int) Math.sqrt(n);
        for (int i = 0; i < n; i++) {
            Set<Character> rowSet = new HashSet<>();
            Set<Character> colSet = new HashSet<>();
            Set<Character> boxSet = new HashSet<>();
            for (int j = 0; j < n; j++) {
                if (board[i][j] != '.' && !rowSet.add(board[i][j])) {
                    return false;
                }
                if (board[j][i] != '.' && !colSet.add(board[j][i])) {
                    return false;
                }
                int row = (i / m) * m + j / m;
                int col = (i % m) * m + j % m;
                if (board[row][col] != '.' && !boxSet.add(board[row][col])) {
                    return false;
                }
            }
        }
        return true;
    }
}

File: src/main/python/36_valid_sudoku/set_solution.py
from math import sqrt
from typing import List

from task import Task


class SetSolution(Task):
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        n = len(board)
        size = int(sqrt(n))
        for i in range(n):
            seen_col = set()
            seen_row = set()
            seen_box = set()
            for j in range(n):
                if board[i][j] in seen_col:
                    return False
                elif board[i][j] != '.':
                    seen_col.add(board[i][j])
                if board[j][i] in seen_row:
                    return False
                elif board[j][i] != '.':
                    seen_row.add(board[j][i])
                row = size * (i // size) + j // size
                col = size * (i % size) + j % size
                if board[row][col] in seen_box:
                    return False
                elif board[row][col] != '.':
                    seen_box.add(board[row][col])
        return True
